File: pset6/readability/readability.py
from cs50 import get_string
import re

text = get_string("Text: ")

sentences = re.split('[.!?] ', text)
sentencecount = len(sentences)

words = []
for sentence in sentences:
    words += sentence.split()
wordcount = len(words)

lettercount = 0;
for word in words:
    for letter in word:
        if letter >= 'a' and letter <= 'z':
            lettercount += 1
        elif letter >= 'A' and letter <= 'Z':
            lettercount += 1

# Coleman-Liau index: 0.0588 * L - 0.296 * S - 15.8 where
# L is the average number of letters per 100 words and
# S is the average number of sentences per 100 words
let = 100 * lettercount / wordcount
sen = 100 * sentencecount / wordcount
cole_liau = 0.0588 * let - 0.296 * sen - 15.8

grade = "Grade "
if cole_liau >= 16:
    grade += "16+"
elif cole_liau < 1:
    grade = "Before Grade 1"
else:
    grade += str(round(cole_liau))

print(grade)

File: pset2/readability/readability.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <math.h>

int count_letters(string s);
int count_words(string s);
int count_sentences(string s);

int main(void)
{
    // get input from the user
    string text = get_string("Text: ");

    // Use the Coleman-Liau formula and round to the nearest integer
    float letterAvg = (float) count_letters(text) / (float) count_words(text) * 100;
    float sentenceAvg = (float) count_sentences(text) / (float) count_words(text) * 100;
    float index = 0.0588 * letterAvg - 0.296 * sentenceAvg - 15.8;
    int grade = round(index);

    // print "Before Grade 1", "Grade X", or "Grade 16+"
    if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}

int count_letters(string s)
{
    int letters = 0;
    for (int i = 0, length = strlen(s); i < length; i++)
    {
        // if the ASCII number at s[i] maps to a letter, increment letters
        if ((s[i] >= 65 && s[i] <= 90) || (s[i] >= 97 && s[i] <= 122))
        {
            letters++;
        }
    }
    return letters;
}

int count_words(string s)
{
    int words = 0;
    for (int i = 0, length = strlen(s); i < length; i++)
    {
        // If the current character is a ' ' or end-of-sentence punctuation,
        // and the previous character is NOT end-of-sentence punctuation
        // (so as not to double count)...increment words
        if (s[i] == ' ' || s[i] == '.' || s[i] == '!' || s[i] == '?')
        {
            if (s[i - 1] != '.' && s[i - 1] != '!' && s[i - 1] != '?')
            {
                words++;
            }
        }
    }
    return words;
}

int count_sentences(string s)
{
    int sentences = 0;

    // if s[i] contains end-of-sentence punctuation, count it as a sentence
    for (int i = 0, length = strlen(s); i < length; i++)
    {
        if (s[i] == '.' || s[i] == '!' || s[i] == '?')
        {
            sentences++;
        }
    }

    return sentences;
}